File: app/src/test/java/com/przemolab/oknotifier/models/ContestTests.java
package com.przemolab.oknotifier.models;

import com.przemolab.oknotifier.utils.DateUtils;

import org.junit.Test;

import java.util.Date;

import static junit.framework.Assert.assertEquals;

public class ContestTests {

    @Test
    public void constructor_setsAllFields() {
        // when
        Contest contest = createContest();

        // then
        assertEquals("abc", contest.getId());
        assertEquals("name", contest.getName());
    }

    @Test
    public void setId_setsId() {
        // given
        Contest contest = createContest();

        // when
        contest.setId("changed id");

        // then
        assertEquals("changed id", contest.getId());
    }

    @Test
    public void setName_setsName() {
        // given
        Contest contest = createContest();

        // when
        contest.setName("changed name");

        // then
        assertEquals("changed name", contest.getName());
    }

    @Test
    public void setStartDate_setsStartDate() {
        // given
        Contest contest = createContest();
        Date newDate = DateUtils.getDate(2000, 1, 1, 10, 30);

        // when
        contest.setStartDate(newDate);

        // then
        assertEquals(newDate, contest.getStartDate());
    }

    @Test
    public void setEndDate_setsEndDate() {
        // given
        Contest contest = createContest();
        Date newDate = DateUtils.getDate(2000, 1, 1, 10, 30);

        // when
        contest.setEndDate(newDate);

        // then
        assertEquals(newDate, contest.getEndDate());
    }

    @Test
    public void setNumberOfContestants_setsNumberOfContestants() {
        // given
        Contest contest = createContest();

        // when
        contest.setNumberOfContestants(100);

        // then
        assertEquals(100, contest.getNumberOfContestants());
    }

    @Test
    public void setNumberOfProblems_setsNumberOfProblems() {
        // given
        Contest contest = createContest();

        // when
        contest.setNumberOfProblems(100);

        // then
        assertEquals(100, contest.getNumberOfProblems());
    }

    private Contest createContest() {
        String id = "abc";
        String name = "name";
        Date startDate = DateUtils.getDate(2015, 6, 10, 16, 0);
        Date endDate = DateUtils.getDate(2015, 6, 15, 16, 0);
        int numberOfContestants = 5;
        int numberOfProblems = 10;

        return new Contest(id, name, startDate, endDate, numberOfContestants, numberOfProblems);
    }
}

File: app/src/test/java/com/przemolab/oknotifier/models/ContestTests.kt
package com.przemolab.oknotifier.models

import com.przemolab.oknotifier.utils.DateUtils

import org.junit.Test

import java.util.Date

import junit.framework.Assert.assertEquals

class ContestTests {

    @Test
    fun constructor_setsAllFields() {
        // when
        val contest = createContest()

        // then
        assertEquals("abc", contest.contestId)
        assertEquals("name", contest.name)
        assertEquals("6/10/15 4:00PM", contest.startDateFormatted)
        assertEquals("6/15/15 5:00PM", contest.endDateFormatted)
        assertEquals(5, contest.numberOfContestants)
        assertEquals(10, contest.numberOfProblems)
    }

    @Test
    fun getStartDateFormatted_returnsStartDateInExpectedFormat() {
        // given
        val contest = createContest()

        // when
        val result = contest.startDateFormatted

        // then
        assertEquals("6/10/15 4:00PM", result)
    }

    @Test
    fun getEndDateFormatted_returnsEndDateInExpectedFormat() {
        // given
        val contest = createContest()

        // when
        val result = contest.endDateFormatted

        // then
        assertEquals("6/15/15 5:00PM", result)
    }

    @Test
    fun setId_setsId() {
        // given
        val contest = createContest()

        // when
        contest.contestId = "changed id"

        // then
        assertEquals("changed id", contest.contestId)
    }

    @Test
    fun setName_setsName() {
        // given
        val contest = createContest()

        // when
        contest.name = "changed name"

        // then
        assertEquals("changed name", contest.name)
    }

    @Test
    fun setStartDate_setsStartDate() {
        // given
        val contest = createContest()
        val newDate = DateUtils.getDate(2000, 1, 1, 10, 30, 0)

        // when
        contest.startDate = newDate

        // then
        assertEquals(newDate, contest.startDate)
    }

    @Test
    fun setEndDate_setsEndDate() {
        // given
        val contest = createContest()
        val newDate = DateUtils.getDate(2000, 1, 1, 10, 30, 0)

        // when
        contest.endDate = newDate

        // then
        assertEquals(newDate, contest.endDate)
    }

    @Test
    fun setNumberOfContestants_setsNumberOfContestants() {
        // given
        val contest = createContest()

        // when
        contest.numberOfContestants = 100

        // then
        assertEquals(100, contest.numberOfContestants)
    }

    @Test
    fun setNumberOfProblems_setsNumberOfProblems() {
        // given
        val contest = createContest()

        // when
        contest.numberOfProblems = 100

        // then
        assertEquals(100, contest.numberOfProblems)
    }

    @Test
    fun setSubscribed_setsSubscribed() {
        // given
        val contest = createContest()

        // when
        contest.isSubscribed = true

        // then
        assertEquals(true, contest.isSubscribed)
    }

    private fun createContest(): Contest {
        val id = "abc"
        val name = "name"
        val startDate = DateUtils.getDate(2015, 6, 10, 16, 0, 0)
        val endDate = DateUtils.getDate(2015, 6, 15, 17, 0, 0)
        val numberOfContestants = 5
        val numberOfProblems = 10

        return Contest(id, name, startDate, endDate, numberOfContestants, numberOfProblems)
    }
}